src/pathfinder/opf/opf_ship.cpp

```diff
@@ -220,3 +227,57 @@ Track OPFShipChooseTrack(const Ship *v, TileIndex tile, DiagDirection enterdir,
 	if (dist <= rev_dist) return track;
 	return INVALID_TRACK; // We could better reverse
 }
+
+FindDepotData OPFShipFindNearestDepot(const Ship *v, int max_penalty)
+{
+	FindDepotData fdd;
+
+	Trackdir trackdir = v->GetVehicleTrackdir();
+
+	/* Argument values for FindShipTrack below, for the current ship direction */
+	DiagDirection enterdir = TrackdirToExitdir(trackdir);
+	TileIndex tile = TILE_ADD(v->tile, TileOffsByDiagDir(enterdir));
+	TrackBits tracks = TrackdirBitsToTrackBits(DiagdirReachesTrackdirs(enterdir));
+
+	/* Argument values for FindShipTrack below, for the reversed ship direction */
+	DiagDirection enterdir_rev = ReverseDiagDir(enterdir);
+	TileIndex tile_rev = v->tile;
+	TrackBits tracks_rev = TrackStatusToTrackBits(GetTileTrackStatus(tile_rev, TRANSPORT_WATER, 0)) & DiagdirReachesTracks(enterdir_rev) & TrackdirBitsToTrackBits(TrackdirToTrackdirBits(trackdir));
+
+	Track track; // initialized, but unused
+	if (max_penalty == 0) max_penalty = OPF_MAX_LENGTH;
```

Comment:
Distance, not penalty.

Reply:
It's not a distance either. It's the maximum number of tiles the ship can hop. It is also used by OPF to stop searching further. It's more like a max pathfinder cost combined with number of tiles travelled.

I am using it inside the FOR_ALL_DEPOTS loop in combination with distance manhathan because  the number of tiles hopped can never be less than the distance manhathan. It is, at best, the same. This will help reduce the number of FindShipTrack calls when iterating over all depots. The result that FindShipTrack gives is also used for the same effect.